packers/jws.go

```diff
@@ -238,77 +227,97 @@ func (p *JWSPacker) Unpack(envelope []byte) (*iden3comm.BasicMessage, error) {
 		return nil, errors.New("from field in did document is required")
 	}
 
-	if msg.From != kid {
-		return nil, errors.New("message signer must be the message sender")
+	parsedKid := strings.Split(kid, "#")
+	if len(parsedKid) < 1 {
+		return nil, errors.New("kid is expected in format of valid did string")
+	}
+	if kid != "" && parsedKid[0] != msg.From {
+		return nil, errors.New("sender must use equal kid that contains did identifier")
 	}
 
 	didDoc, err := p.didResolverHandler.Resolve(msg.From)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 
-	vm, err := lookupForKid(didDoc, kid)
-	if err != nil {
-		return nil, err
-	}
-
-	err = checkAlgorithmSupport(alg, vm)
+	vms, err := resolveVerificationMethods(didDoc)
 	if err != nil {
 		return nil, err
 	}
 
-	if vm.BlockchainAccountID == "" && vm.EthereumAddress == "" {
-		wk, err := extractVerificationKey(alg, vm)
+	if kid != "" {
+		var vm verifiable.CommonVerificationMethod
+		vm, err = findVerificationMethodByID(vms, kid)
 		if err != nil {
 			return nil, err
 		}
-		_, err = jws.Verify(envelope, wk)
-		if err != nil {
-			return nil, errors.WithStack(err)
-		}
-	} else {
-		base64Token, err := jws.Compact(token)
-		if err != nil {
-			return nil, errors.WithStack(err)
-		}
-		base64TokenParts := strings.Split(string(base64Token), ".")
-		signedData := base64TokenParts[0] + "." + base64TokenParts[1]
-		hash := sha256.Sum256([]byte(signedData))
-		sig := token.Signatures()[0].Signature()
-		recoveredKey, err := ecc.RecoverEthereum(hash[:], sig)
-		if err != nil {
-			return nil, errors.WithStack(err)
-		}
+		vms = []verifiable.CommonVerificationMethod{vm}
+	}
 
-		ethAddress := "0x" + hex.EncodeToString(keccak256.Hash(recoveredKey[1:])[12:])
+	for i := range vms {
 
-		if vm.EthereumAddress != "" {
-			if !strings.EqualFold(ethAddress, vm.EthereumAddress) {
-				return nil, errors.New("invalid signature from ethereum address")
+		err = checkAlgorithmSupport(alg, vms[i])
+		if err != nil {
+			// skip verification method check if algorithm is not supported
+			continue
+		}
+		if vms[i].BlockchainAccountID == "" && vms[i].EthereumAddress == "" {
+			wk, err := extractVerificationKey(alg, vms[i])
+			if err != nil {
+				continue
+			}
+			_, err = jws.Verify(envelope, wk)
+			if err != nil {
+				continue
 			}
 		} else {
-			blockchainAccountIDParts := strings.Split(vm.BlockchainAccountID, ":")
-			address := blockchainAccountIDParts[len(blockchainAccountIDParts)-1]
+			base64Token, err := jws.Compact(token)
+			if err != nil {
+				continue
+			}
+			base64TokenParts := strings.Split(string(base64Token), ".")
+			signedData := base64TokenParts[0] + "." + base64TokenParts[1]
```

Comment:
Panic is possible here. Please, check len to be sure that `base64TokenParts` variable has needed elements.